### PR TITLE
Backend/#10/database api

### PR DESCRIPTION
**Pull request by ecl1ipse**
Created dbCreate, dbRead, dbUpdate and dbDel functions in db_api.ts.

### CODE REVIEW

File: server/db.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { Env, config } from './config'
+const settings = config[process.env.NODE_ENV || Env.DEV]
+import Database from './dbApi'
+
+const test = async () => {
```
Comment (Mysterise) [inline server/db.ts]: Will be moving this stuff out in #16

File: server/dbApi.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import { MongoClient, Db } from 'mongodb'
+import { Env, config } from './config'
+
+const debug = true
+
+// URL for database
+//const env: Env = Env[process.env.NODE_ENV] || Env.development
```
Comment (Mysterise) [inline server/dbApi.ts]: Remove outdated commented line

File: server/dbApi.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+  }
+
+  getDb = async (): Promise<Db> => {
+    if (!this.client) await this.connect()
+    const db = this.client.db(dbName)
+    return db
+  }
+
+  dbCreate = async doc => {
```
Comment (Mysterise) [inline server/dbApi.ts]: @Skithy do we want to be consistent in function parameters, and enforce parenthesis?

File: server/db.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { Env, config } from './config'
+const settings = config[process.env.NODE_ENV || Env.DEV]
+import Database from './dbApi'
```
Comment (Mysterise) [inline server/db.ts]: shouldn't mix import lines with declarations

File: server/db.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+const test = async () => {
+  console.log('hai')
+  await Database.dbCreate({ id: 'COMP1521', vac: '0', gh: 8 })
+  var doc = await Database.dbRead('COMP1521')
```
Comment (Mysterise) [inline server/db.ts]: Should be const?

File: server/dbApi.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+// URL for database
+//const env: Env = Env[process.env.NODE_ENV] || Env.development
+const url = config[process.env.NODE_ENV || Env.DEV].database
+
+// DB Name
+const dbName = 'Notangles'
+const colName = 'Courses'
+
+// Create a the MongoClient
```
Comment (Mysterise) [inline server/dbApi.ts]: Wrong comment/misplaced? Should be a comment describing the Database class instead.

File: server/dbApi.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    var doc = await col.findOne({ id: id })
+    if (debug) console.log('Reading doc: ' + doc)
+    return doc
+  }
+
+  dbUpdate = async (id, doc) => {
+    const db = await this.getDb()
+    const col = db.collection(colName)
+    try {
+      await col.updateOne({ id: id }, { $set: doc })
```
Comment (Mysterise) [inline server/dbApi.ts]: Might be good to have a small comment describing why $set is here for future ref

File: server/app.ts [hunk 8]
```diff
@@ -1,11 +1,13 @@
+import db from './db_api'
+
 const express = require('express')
```
Comment (Skithy) [inline server/app.ts]: convert this one to import as well. try to not mix up imports and requires, keep it consistent

File: server/config.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import * as secret from './secret.json'
```
Comment (Skithy) [inline server/config.ts]: i think, should add a comment here that we should make the secret.json if it you currently don't have one

File: server/db.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { Env, config } from './config'
+const settings = config[process.env.NODE_ENV || Env.DEV]
+import Database from './dbApi'
+
```
Comment (Skithy) [inline server/db.ts]: have all the imports at the top, then do stuff after
```
import X from ''
import Y from ''

const z = ...
```

File: server/config.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  },
+
+  [Env.STAGING]: {
+    database: secret.staging,
+  },
+
+  [Env.PROD]: {
+    database: secret.prod,
+  },
+}
```
Comment (Skithy) [inline server/config.ts]: actually, you can do something like
```
const allConfig: Record<Env, IConfig> = {
  [Env.DEV]: {
    database: secret.dev,
  },

  [Env.STAGING]: {
    database: secret.staging,
  },

  [Env.PROD]: {
    database: secret.prod,
  },
}

export const config = allConfig[process.env.NODE_ENV || Env.DEV]
```
this way, you don't have to check env every time you wanted to use config

File: server/dbApi.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
+import { MongoClient, Db } from 'mongodb'
+import { Env, config } from './config'
+
+const debug = true
```
Comment (Skithy) [inline server/dbApi.ts]: maybe instead have this in the config instead

File: server/dbApi.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    const db = await this.getDb()
+    const col = db.collection(colName)
+    await col.insertOne(doc)
+    if (debug) console.log('Created doc: ' + doc)
+  }
+
+  dbRead = async id => {
+    const db = await this.getDb()
+    const col = db.collection(colName)
+    var doc = await col.findOne({ id: id })
```
Comment (Skithy) [inline server/dbApi.ts]: try not to use var (var actually works differently from let and const, which can lead to some confusing behaviour. if you want, you can research about scopes in javascript and all the quirks of it)
always use const, then let

Comment (Skithy) [reply server/dbApi.ts]: for an object, if the key is the same name as the value variable, you can exclude it
```
col.findOne({ id })
```

File: server/dbApi.ts (hunk 3 above)
Comment (Skithy) [inline server/dbApi.ts]: maybe dbInsert or dbAdd? dbCreate sounds like you are trying to make a database.

File: server/dbApi.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  dbCreate = async doc => {
+    const db = await this.getDb()
+    const col = db.collection(colName)
+    await col.insertOne(doc)
+    if (debug) console.log('Created doc: ' + doc)
+  }
+
+  dbRead = async id => {
```
Comment (Skithy) [inline server/dbApi.ts]: i think here, it is good to specify what data-type id should be.
```
dbRead = async (id: string) => {
```
or
```
dbRead = async (id: number) => {
```

File: server/dbApi.ts (hunk 7 above)
Comment (Skithy) [inline server/dbApi.ts]: same here `col.updateOne({ id }, { $set: doc })`

File: server/dbApi.ts (hunk 3 above)
Comment (Skithy) [reply server/dbApi.ts]: should be a in-built prettier rule
i think its okay without params for single, looks a bit cleaner
```
const cats = animals.map(animal => animal.type === 'cat')
```
```
const cats = animals.map((animal) => animal.type === 'cat')
```

File: server/dbApi.ts (hunk 7 above)
Comment (Skithy) [reply server/dbApi.ts]: i think it's better to just refer to the official documentation here, and searching yourself how does updateOne works.

Comment (Skithy) [reply server/dbApi.ts]: Otherwise we end up with tons of comments about third party library stuff.

Comment (Mysterise) [reply server/dbApi.ts]: Agreed, though think this is an exception - most library stuff are fairly intuitive, whereas the $set syntax looks alien with the rest of the codebase so far

**Pull request merged**